cmd/transcriber/call/live_captions.go

```diff
@@ -0,0 +1,461 @@
+package call
+
+import (
+	"errors"
+	"fmt"
+	"github.com/mattermost/calls-transcriber/cmd/transcriber/apis/whisper.cpp"
+	"github.com/mattermost/calls-transcriber/cmd/transcriber/config"
+	"github.com/mattermost/calls-transcriber/cmd/transcriber/opus"
+	"github.com/mattermost/calls-transcriber/cmd/transcriber/transcribe"
+	"github.com/mattermost/mattermost-plugin-calls/server/public"
+	"github.com/streamer45/silero-vad-go/speech"
+	"log/slog"
+	"path/filepath"
+	"time"
+)
+
+const (
+	transcriberQueueChBuffer = 1
+	tickRate                 = 2 * time.Second
+	maxWindowSize            = 8 * time.Second
+	windowPressureLimitSec   = 12                                                           // at this point cut the audio down to prevent a death spiral
+	pktPayloadChBuffer       = trackInAudioRate / trackInFrameSize * windowPressureLimitSec // hard drop after windowPressureLimitSec seconds of audio backing up
+	removeWindowAfterSilence = 3 * time.Second
+
+	// VAD settings
+	vadWindowSizeInSamples  = 512
+	vadThreshold            = 0.5
+	vadMinSilenceDurationMs = 350
+	vadSpeechPadMs          = 200
+	minSpeechLengthSamples  = 1000 * trackOutAudioSamplesPerMs // 1 second of speech
+)
+
+type captionPackage struct {
+	pcm   []float32
+	retCh chan string
+}
+
+func (t *Transcriber) processLiveCaptionsForTrack(ctx trackContext, pktPayloads <-chan []byte, doneCh <-chan struct{}) {
+	opusDec, err := opus.NewDecoder(trackOutAudioRate, trackAudioChannels)
+	if err != nil {
+		slog.Error("processLiveCaptionsForTrack: failed to create opus decoder for live captions",
+			slog.String("err", err.Error()), slog.String("trackID", ctx.trackID))
+		return
+	}
+	defer func() {
+		if err := opusDec.Destroy(); err != nil {
+			slog.Error("processLiveCaptionsForTrack: failed to destroy decoder", slog.String("err", err.Error()),
+				slog.String("trackID", ctx.trackID))
+		}
+	}()
+
+	// Setup the VAD
+	sd, err := speech.NewDetector(speech.DetectorConfig{
+		ModelPath:  filepath.Join(getModelsDir(), "silero_vad.onnx"),
+		SampleRate: trackOutAudioRate,
+
+		// set WindowSize to 512 to get as fine-grained detection as possible (for when
+		// the number of samples don't cleanly divide into the WindowSize
```

Comment:
Is this still the case? Have we tested with the larger `1536`?

Reply:
I did test just now, I don't think I can tell much of a difference, so I'd like to stay with 512 since it's what silero-vad seems to have been trained with (its examples and quality metrics pages).